src/Grpc.Net.Client/Internal/Retry/HedgingCall.cs

```diff
@@ -0,0 +1,410 @@
+                }
+                else if (_activeCalls.Count == 1 && (Channel.RetryThrottling?.IsRetryThrottlingActive() ?? false))
+                {
+                    // This is the last active call and throttling is active.
+                    CommitCall(call, CommitReason.Throttled);
+                }
+                else
+                {
+                    // Call isn't used and can be cancelled.
+                    // Note that the call could have already been removed and disposed if the
+                    // hedging call has been finalized or disposed.
+                    if (_activeCalls.Remove(call))
+                    {
+                        call.Dispose();
+                    }
+                }
+            }
+        }
+
+        protected override void OnCommitCall(IGrpcCall<TRequest, TResponse> call)
+        {
+            _activeCalls.Remove(call);
+
+            CleanUpUnsynchronized();
+        }
+
+        private void CleanUpUnsynchronized()
+        {
+            while (_activeCalls.Count > 0)
+            {
+                _activeCalls[_activeCalls.Count - 1].Dispose();
+                _activeCalls.RemoveAt(_activeCalls.Count - 1);
+            }
+        }
+
+        protected override void StartCore(Action<GrpcCall<TRequest, TResponse>> startCallFunc)
+        {
+            var hedgingDelay = _hedgingPolicy.HedgingDelay.GetValueOrDefault();
+            if (hedgingDelay == TimeSpan.Zero)
+            {
+                // If there is no delay then start all call immediately
+                while (_callsAttempted < MaxRetryAttempts)
+                {
+                    _ = StartCall(startCallFunc);
+
+                    // Don't send additional calls if retry throttling is active.
+                    if (Channel.RetryThrottling?.IsRetryThrottlingActive() ?? false)
+                    {
+                        Log.AdditionalCallsBlockedByRetryThrottling(Logger);
+                        break;
+                    }
+
+                    lock (Lock)
+                    {
+                        // Don't send additional calls if call has been commited.
+                        if (CommitedCallTask.IsCompletedSuccessfully())
+                        {
+                            break;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                CreateHedgingCallsTask = CreateHedgingCalls(startCallFunc);
+            }
+        }
+
+        private async Task CreateHedgingCalls(Action<GrpcCall<TRequest, TResponse>> startCallFunc)
+        {
+            Log.StartingRetryWorker(Logger);
+
+            try
+            {
+                var hedgingDelay = _hedgingPolicy.HedgingDelay.GetValueOrDefault();
+
+                while (_callsAttempted < MaxRetryAttempts)
+                {
+                    _ = StartCall(startCallFunc);
+
+                    await HedgingDelayAsync(hedgingDelay).ConfigureAwait(false);
```

Comment:
"If a non-fatal status code is received from a hedged request, then the next hedged request in line is sent immediately, shortcutting its hedging delay."
https://github.com/grpc/proposal/blob/8a695490fed36c59d2974f1bb38ab399447769fb/A6-client-retries.md#hedging-policy

Where would that be handled?

Reply:
Hmm, I think I missed that piece of logic.

Reply:
Done.

Right now a TCS combined with WhenAny is being used to escape the `Task.Delay` here. I'm pretty sure there are performance reasons why not canceling a Task.Delay you no longer care about is bad.

Double checking: I should use a CTS instead, right?

Reply:
Updated so Task.Delay is canceled if not completed.